### PR TITLE
lookup secrets only by key

### PR DESCRIPTION
**Pull request by grosser**
solving a few issues:
 - being able to look up production issues in test env
 - needing to hardcode in which env/group/project their secret is
 - being unable to add a specific secret to a deploy-group or project without changing the projects kubernetes file

Implementation details:
 - bar -> find all possible keys (global/global/global/bar, production/global/global/bar etc) -> pick the most specific one
 - showing the users the exact keys in the secrets UI so it's easier to understand
 - adding read_multi for efficient multi-key reads

Followup work:
 - extract resolver to separate file and move the tests that are spread between template and executor there

screen shot 2016-08-02 at 4 29 01 pm

@irwaters @jonmoter 

this will break all existing secrets when deploying ... so needs some manual hand-holding/fixing

### CODE REVIEW

File: app/models/secret_storage.rb [hunk 1]
```diff
@@ -63,13 +76,22 @@ class << self
       def read(key)
         key = vault_path(key)
         result = vault_client.read(key)
-        raise ActiveRecord::RecordNotFound if result.data[:vault].nil?
+        return if result.data[:vault].nil?
```
Comment (grosser) [inline app/models/secret_storage.rb]: no need to raise here since SeccretStorage will do that

**Comment by jonmoter** [comment]
👍 on 💚

**Comment by grosser** [comment]
added full key as hover effect

**Comment by grosser** [comment]
deployed kube-watcher https://samsontest.zende.sk/projects/kube_service_watcher/deploys/2766 with new secrets https://github.com/zendesk/kube_service_watcher/pull/16

**Comment by grosser** [comment]
once this hits production these secrets need to be updated:

https://samson.zende.sk/admin/commands?utf8=%E2%9C%93&search%5Bquery%5D=secret%3A%2F%2F&search%5Bproject_id%5D=&commit=Search

**Comment by vitalis** [comment]
@grosser I deployed this commit to our servers and the deploy doesn't work anymore, can you post the required changes for production? which secrets need to be updated?

**Comment by grosser** [comment]
go to 
/admin/commands?utf8=%E2%9C%93&search%5Bquery%5D=secret%3A%2F%2F&search%5Bproject_id%5D=&commit=Search

and update everything that has the old secret://a/b/c/d structure to secret://d

**Comment by grosser** [comment]
sorry for the breakage ... did not think anyone else was using secrets yet :D

**Comment by vitalis** [comment]
@grosser 
Still no luck, the deploy process just doesn't start, in logs I wee some authentication issue:
I'm attaching screenshot and request logs

screenshot_09_08_2016__1_38_pm

```
2016-08-09T10:35:57.908743+00:00 app[web.1]: Started POST "/projects/apprime-reactor/stages/staging-ae90d866/deploys?deploy%5Breference%5D=master" for 84.108.167.158 at 2016-08-09 10:35:57 +0000
2016-08-09T10:35:57.911026+00:00 app[web.1]: Processing by DeploysController#create as HTML
2016-08-09T10:35:57.911065+00:00 app[web.1]:   Parameters: {"authenticity_token"=>"lqC/1H/t89nvMIOxEjvAXvEfpWH6LqVZ+2mmy3a/1BQjsWvlHxsdq50R6XHGYtiQUpmjbVOkUdYZ1q0z77DzcA==", "deploy"=>{"reference"=>"master"}, "project_id"=>"apprime-reactor", "stage_id"=>"staging-ae90d866"}
2016-08-09T10:35:57.970983+00:00 app[web.1]: get https://api.github.com/repos/applicaster/apprime-reactor/branches/master
2016-08-09T10:35:58.049050+00:00 app[web.1]: 304
2016-08-09T10:35:58.053772+00:00 app[web.1]:   Rendered plugins/slack_webhooks/app/views/samson_slack_webhooks/notification.text.erb (89.3ms)
2016-08-09T10:35:58.245185+00:00 app[web.1]: Creating GitHub Deployment...
2016-08-09T10:35:58.247897+00:00 app[web.1]: post https://api.github.com/repos/applicaster/apprime-reactor/deployments
2016-08-09T10:35:58.299479+00:00 app[web.1]: 201
2016-08-09T10:35:58.319330+00:00 app[web.1]: Redirected to https://apprime-samson.herokuapp.com/projects/apprime-reactor/deploys/177
2016-08-09T10:35:58.319497+00:00 app[web.1]: Completed 302 Found in 408ms (ActiveRecord: 20.6ms)
2016-08-09T10:35:58.345800+00:00 app[web.1]:   Rendered plugins/slack_webhooks/app/views/samson_slack_webhooks/notification.text.erb (2.6ms)
2016-08-09T10:35:58.481204+00:00 app[web.1]: Started GET "/projects/apprime-reactor/deploys/177" for 84.108.167.158 at 2016-08-09 10:35:58 +0000
2016-08-09T10:35:58.483769+00:00 app[web.1]: Processing by DeploysController#show as HTML
2016-08-09T10:35:58.483793+00:00 app[web.1]:   Parameters: {"project_id"=>"apprime-reactor", "id"=>"177"}
2016-08-09T10:35:58.329020+00:00 heroku[router]: at=info method=POST path="/projects/apprime-reactor/stages/staging-ae90d866/deploys?deploy%5Breference%5D=master" host=apprime-samson.herokuapp.com request_id=f046aa07-49fe-4443-bf20-3baecb83c632 fwd="84.108.167.158" dyno=web.1 connect=1ms service=414ms status=302 bytes=994
2016-08-09T10:35:58.536568+00:00 heroku[router]: at=info method=GET path="/projects/apprime-reactor/deploys/177" host=apprime-samson.herokuapp.com request_id=0978e9f9-7f86-4bc3-97e9-49d208356b80 fwd="84.108.167.158" dyno=web.1 connect=1ms service=48ms status=200 bytes=8716
2016-08-09T10:35:58.503650+00:00 app[web.1]: Updating GitHub Deployment Status...
2016-08-09T10:35:58.508727+00:00 app[web.1]: get https://api.github.com/repos/applicaster/apprime-reactor/deployments/8430387
2016-08-09T10:35:58.513283+00:00 app[web.1]:   Rendered deploys/_header.html.erb (2.9ms)
2016-08-09T10:35:58.513625+00:00 app[web.1]:   Rendered plugins/new_relic/app/views/samson_new_relic/_deploy_tab_nav.html.erb (0.0ms)
2016-08-09T10:35:58.513983+00:00 app[web.1]:   Rendered plugins/slack_webhooks/app/views/samson_slack_webhooks/_notify_buddy_box.html.erb (0.0ms)
2016-08-09T10:35:58.514589+00:00 app[web.1]:   Rendered jobs/_log.html.erb (0.1ms)
2016-08-09T10:35:58.514684+00:00 app[web.1]:   Rendered shared/_output.html.erb (0.6ms)
2016-08-09T10:35:58.514810+00:00 app[web.1]:   Rendered plugins/new_relic/app/views/samson_new_relic/_deploy_tab_body.html.erb (0.0ms)
2016-08-09T10:35:58.515027+00:00 app[web.1]:   Rendered deploys/show.html.erb within layouts/application (21.1ms)
2016-08-09T10:35:58.525315+00:00 app[web.1]:   Rendered shared/_header.html.erb (9.3ms)
2016-08-09T10:35:58.525533+00:00 app[web.1]:   Rendered shared/_alerts.html.erb (0.1ms)
2016-08-09T10:35:58.526206+00:00 app[web.1]:   Rendered shared/_footer.html.erb (0.0ms)
2016-08-09T10:35:58.526405+00:00 app[web.1]: Completed 200 OK in 43ms (Views: 18.0ms | ActiveRecord: 20.5ms)
2016-08-09T10:35:58.543494+00:00 app[web.1]: 200
2016-08-09T10:35:58.548360+00:00 app[web.1]: post https://api.github.com/repos/applicaster/apprime-reactor/deployments/8430387/statuses
2016-08-09T10:35:58.590923+00:00 app[web.1]: 201
2016-08-09T10:35:58.626759+00:00 app[web.1]: Started GET "/projects/apprime-reactor/deploys/177" for 54.89.92.4 at 2016-08-09 10:35:58 +0000
2016-08-09T10:35:58.628605+00:00 app[web.1]: Processing by DeploysController#show as */*
2016-08-09T10:35:58.628628+00:00 app[web.1]:   Parameters: {"project_id"=>"apprime-reactor", "id"=>"177"}
2016-08-09T10:35:58.628998+00:00 app[web.1]: Halted as unauthorized! threw :warden
2016-08-09T10:35:58.629093+00:00 app[web.1]: Completed   in 0ms (ActiveRecord: 0.0ms)
2016-08-09T10:35:58.646991+00:00 app[web.1]: Started GET "/login?origin=%2Fprojects%2Fapprime-reactor%2Fdeploys%2F177" for 54.89.92.4 at 2016-08-09 10:35:58 +0000
2016-08-09T10:35:58.649197+00:00 app[web.1]: Processing by SessionsController#new as */*
2016-08-09T10:35:58.649219+00:00 app[web.1]:   Parameters: {"origin"=>"/projects/apprime-reactor/deploys/177"}
2016-08-09T10:35:58.650517+00:00 app[web.1]:   Rendered sessions/new.html.erb within layouts/application (0.4ms)
2016-08-09T10:35:58.652161+00:00 app[web.1]:   Rendered shared/_header.html.erb (0.5ms)
2016-08-09T10:35:58.652328+00:00 app[web.1]:   Rendered shared/_alerts.html.erb (0.1ms)
2016-08-09T10:35:58.653100+00:00 app[web.1]:   Rendered shared/_footer.html.erb (0.0ms)
2016-08-09T10:35:58.653285+00:00 app[web.1]: Completed 200 OK in 4ms (Views: 3.7ms | ActiveRecord: 0.0ms)
2016-08-09T10:35:58.681921+00:00 app[web.1]: Started GET "/projects/apprime-reactor/deploys/177" for 52.90.146.42 at 2016-08-09 10:35:58 +0000
2016-08-09T10:35:58.683779+00:00 app[web.1]: Processing by DeploysController#show as */*
2016-08-09T10:35:58.683834+00:00 app[web.1]:   Parameters: {"project_id"=>"apprime-reactor", "id"=>"177"}
2016-08-09T10:35:58.684195+00:00 app[web.1]: Halted as unauthorized! threw :warden
2016-08-09T10:35:58.684227+00:00 app[web.1]: Completed   in 0ms (ActiveRecord: 0.0ms)
2016-08-09T10:35:58.691860+00:00 app[web.1]: Started GET "/login?origin=%2Fprojects%2Fapprime-reactor%2Fdeploys%2F177" for 52.90.146.42 at 2016-08-09 10:35:58 +0000
2016-08-09T10:35:58.693539+00:00 app[web.1]: Processing by SessionsController#new as */*
2016-08-09T10:35:58.693562+00:00 app[web.1]:   Parameters: {"origin"=>"/projects/apprime-reactor/deploys/177"}
2016-08-09T10:35:58.694588+00:00 app[web.1]:   Rendered sessions/new.html.erb within layouts/application (0.4ms)
2016-08-09T10:35:58.695834+00:00 app[web.1]:   Rendered shared/_header.html.erb (0.3ms)
2016-08-09T10:35:58.695996+00:00 app[web.1]:   Rendered shared/_alerts.html.erb (0.1ms)
2016-08-09T10:35:58.696882+00:00 app[web.1]:   Rendered shared/_footer.html.erb (0.0ms)
2016-08-09T10:35:58.697054+00:00 app[web.1]: Completed 200 OK in 3ms (Views: 3.1ms | ActiveRecord: 0.0ms)
```

**Comment by vitalis** [comment]
Also if I roll back to `v822` everything works fine

**Comment by grosser** [comment]
might be unrelated to the secrets change ... `DeploysController#show` -> unauthorized seems weird ... maybe the deploy failed and never got started and that somehow results in an unauthorized :/

**Comment by grosser** [comment]
hmm no ... you are already on the deploy page ... so weird

**Comment by grosser** [comment]
hmm I think we need to add request-id to the stdout logs ... that will happen with rails 5 upgrade anyway I think

**Comment by grosser** [comment]
set `config.log_tags = [ :uuid ]` to see what request exactly fails ... then maybe reproduce it using browser or curl to find the root cause ... make sure to replace all old `secret://` commands ... just in case one slipped by ... or make a new stage that has none of them .... alternatively only revert this commit and see if that also fixes it ...

**Pull request merged**